### ISSUE TITLE
Texture over gauge needle

### ISSUE THREAD

**Issue by DIYLAB-DE** [vindar/tgx#6]
Hi Arvind,

I would like to texture a needle as shown here in the thread and in the video:

https://forum.pjrc.com/threads/59506-3D-Rendering-on-Teensy?p=232392&viewfull=1#post232392
https://youtu.be/NJ7Aby0ym6A

Unfortunately I can't use your 3D part of the TGX library because I have no experience in 3D :/
Can you help me please?

Kind regards
Bruno

**Comment by vindar** [comment]
Hello Bruno, 

Just to let you know that I made some important changes to the library. I did my best to preserve compatibility but I completely cannot guarantee that everything will still compile without trouble because I had to do some important refactoring. 

Here are some new methods that may interest you:

- `blitScaledRotated` : draw a sprite image with a prescribe angle and scale. This method is fast, use bilinear interpolation and sub-pixel precision and blending for very high quality drawing. 

- `drawTexturedTriangle` / `drawTexturedQuad` : blit part of a sprite onto a triangle / quad with bilinear interpolation, sub-pixel precision  and blending. The method is more powerful than `blitScaledRotated` but a bit more difficult to use and require more set up.

I am really pleased with the results ! I think these methods are really powerful and allow to do very fancy graphics. Note in particular that, even if the framebuffer image is 16 bit RGB565 (so no transparency), the blit image can be in any format, for example RGB32 with transparency to perform real blending/masking.

I added an example called 'CrazyClock' to the library. You can check it out to see how to use the new `blitScaledRotated` methods. It is now easy to create very fancy needles :-)

Best,

**Comment by vindar** [comment]
I just made a demo video of the new blitting methods: https://www.youtube.com/watch?v=XxL15cQIPi8

**Comment by DIYLAB-DE** [comment]
Hi Arvind,

I have been able to play a bit - it's a lot of fun :o)
Something I noticed though:

When I create the clock face as RGB32 instead of RGB565 for testing, it sometimes twitches a lot when rotating.
I tried this in portrait format as well as in landscape format - same result.

Attached are my test files: [testfaces.zip](https://github.com/vindar/tgx/files/7101886/testfaces.zip)

Is the Teensy overloaded with RGB32?
Maybe it's also me, because I didn't use your Python scripts to convert the images (I'm too stupid for Python), but I wrote a small tool in C# to convert them. Maybe that's the problem? Actually, I just write each pixel as RGBA in the file in your format.

Please have a look what this could be - thanks.

And one more question: the clock face "green" is transparent - how can I output the transparency on the display? At the moment everything is green up to the edge.

Best regards and I hope I'm not giving you a headache,
Bruno

**Comment by vindar** [comment]
Hello, 

I believe the problem with the twitching is that access to flash memory is really slow. Normally, you do not really notice it because the MPU uses a read cache to speed thing up. However, cache is only good when doing linear access to consecutive memory adresses. Unfortunately, for certain angles, drawing the rotated image will require to access the image pixels in a very unregular pattern (moving a lot around the flash memory) and then the cache become useless and slowdown occur... If that happens, you should try to move the image to a faster memory on the Teensy. Regular RAM (DTCM/ICTM) is fastest, then DMAMEM is also pretty fast, EXTMEM (the external memory if present) is slower and PROGMEM (flash) is slowest... Another, more technical solution is to split the big image in 4 (or even 16 sub image) and blit/rotated each one separately. This will improve coherency and if each sub image be be held in cache, then it will be very fast (notice that the anchor point need not not be inside the image so it not a problem to split them). I will try to put up an example next week when I have some time. 

TLDR; For the time being, just try to remove PROGMEM directive i front of your array (or replace it with DMAMEM) to see if this helps.

Another problem I see with your image is that the RGB32 colors are not in premultiplied color format (see https://en.wikipedia.org/wiki/Alpha_compositing). Every color (r,g,b,a) in your file should be replaced by (r*a / 255 , g*a / 255 , b*a / 2555, a). Note that, for premultiplied alpha colors, there is only 1 transparent color: (0,0,0,0). 

Also, if you want to use a transparent color but do not care about blending really (Ii.e. you just want to create a mask with a given color) then you can indeed use a color type without alpha channel such as RGB565 and then use the method 'blitScaledRotatedMasked' to specify which color should act as the transparent one.

**Comment by DIYLAB-DE** [comment]
Hi Arvind,

exactly, PROGMEM, I should have thought of it myself ;)
Thanks for the quick and interesting answer, I'll try out some things on the weekend.

Kind regards
Bruno

**Comment by DIYLAB-DE** [comment]
Hi Arvind,

cool, it works fine and without twitching - thanks!
Here my little try: [CrazyClock.zip](https://github.com/vindar/tgx/files/7104538/CrazyClock.zip)

Kind regards
Bruno

**Comment by DIYLAB-DE** [comment]
Hi Arvind,

I must be doing something wrong.
Here is a small test program: [NeedleTest.zip](https://github.com/vindar/tgx/files/7109352/NeedleTest.zip)

There I put all buffers and the needle into the fast RAM and do one degree steps.
I just can't get the needle to go faster. That's currently not enough to make fast and smooth movements in the application. And if you look closely, you can see in the movements that the needle partly shows disturbances. This happens both with vsync and without vsync.

Do I have a thinking error or is the method not as fast as desired?

Kind regards
Bruno

**Comment by vindar** [comment]
Hi Bruno, 

I think there are several things at play here. 

1. One should not forget that the ILI9341 has a physical maximum refresh rate around 140Hz, this means that, even if we can draw 1000 frames per second, only 140 will be physically displayed on the screen... For a needle rotating at 1 full turn per second, this means that there is an angle of around 2.6 degree between each draw which is already quite a lot. And when doing 10 turns per seconds (as in the example below), then it is 26 degree between screen refresh !!!

2. If using `tft.setVSyncSpacing(N)`, then the FPS is capped at refreshrate/N. Now, say that N=1 and that the display refresh rate is 140Hz, then calling `tft.update(fb)` 140 times will take 1 second... On the other hand, if vsync is OFF (with `tft.setVSyncSpacing(0)`), then `tft.update(fb)` can be called much more often but some frames will not be displayed or will be mixed with another one (screen tearing). 

3. In your code, I see that you increase the angle by 1 degree at each step. Thus, when vsync is on, you cannot exceed a speed of around 0.3 turn per sec... In fact, is is not a good idea to increase the angle by the same amount at every step of the loop because you do not know how long the drawing or tft.update() may take. Instead, it is much better to compute the current angle of the needle according to the current time (or some other metric that you want to display). This will give you a smoother result.

4. When displaying a needle moving very fast, it might be better to disable vsync to help with aliasing (the stroboscopic effect where the needle appear always at the same angles). Or maybe not, I think it depends on your taste...

5. The edges of the needle in you example are not very smooth. This is because the `blitScaledRotated` method draws the sprite but does not perform antialiasing along the edges of the rotated rectangle. However, the sprite itself is drawn with bilinear filtering so, in order to get clean looking edges, you can simply make the sprite image a little bigger and surround the sprite itself with transparent pixels. This should give you the cleanest result. 

Here is a demo that hopefully will do a better job at explaining the points above. The sprite is in RGB32, it is a red/blue triangle on a transparent background (stored in PROGMEM because it is small enough to hold entirely in the MPU cache). As the demo show, the whole screen can be updated at around 880fps and uploads are fast enough so that no screen tearing occurs when vsync is on.  Here the demo cycles between vsync on and vsync off and different rotation speeds.  At high speed, you start to see the stroboscopic effect come into play which illustrates the physical limitation of the screen.

[NeedleTest.zip](https://github.com/vindar/tgx/files/7110465/NeedleTest.zip)

Hopefully this answers your questions.

**Comment by vindar** [comment]
By the way, I just timed how long it take to draw the needle onto the image: 300 microseconds... So yes, it is fast enough :-)

In fact, in the 880FPS obtained (with vsync off) in the example above, most of the time is spend in tft.update(fb) to compute the diff buffer and set up the DMA chain. The time to draw on the whole image (erasing the image, drawing the needle and the texts and the rectangle) is about 500us.

**Comment by vindar** [comment]
> 
> 
> Hi Arvind,
> 
> cool, it works fine and without twitching - thanks!
> Here my little try: [CrazyClock.zip](https://github.com/vindar/tgx/files/7104538/CrazyClock.zip)
> 
> Video: https://youtu.be/Nxy82OHTjhY
> 
> Kind regards
> Bruno

Very nice ! 
Do you mind if I merge it with my crazy clock demo and add it as an example in the library ?

**Comment by DIYLAB-DE** [comment]
Dear Arvind,

thank you (again) for your great explanations - it helps me a lot!
You already realize, I can also implement bigger projects, but I lack a lot of basics in microcontroller. So I program by feeling ;o)

And of course you can always use everything you need from me without being asked, if you like it.

Kind regards
Bruno

**Comment by vindar** [comment]
Thanks. I understand very well: I am also not very familiar with embedded programming either so I am also learning as I go :-)
